Zuzalu Confessions

veronicaz41 opened the pull request:
Confessions message board

ichub reviewed (commented):
next steps:
- address PR comments
- create a log-in flow which feels really solid that would allow users to see confessions only if they are a zuzalu member

further out next steps:
- move this project to its own repo under the `proofcarryingdata` org
- deploy it on `render.com` alongside our other sites
- style it 
- get people to use it

veronicaz41 commented on apps/confessions-client/.eslintrc.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = {
+  root: true,
+  extends: ["custom"],
```
makes sense. fixed

veronicaz41 commented on apps/confessions-client/pages/index.tsx:
```diff
@@ -0,0 +1,186 @@
+
+  // Listen for PCDs coming back from the Passport popup
+  useEffect(() => {
+    window.addEventListener("message", receiveMessage, false);
```
replaced this code with `usePassportPCD` you added

veronicaz41 commented on apps/confessions-server/src/services/metrics.ts:
removed

veronicaz41 commented on apps/confessions-server/src/services/telemetry.ts:
removed

veronicaz41 commented on apps/confessions-server/src/util/declarations/libhoney.d.ts:
removed
right, these was blindly copied from consumer-server

veronicaz41 commented on apps/confessions-client/CHANGELOG.md:
removed

veronicaz41 commented on apps/confessions-server/CHANGELOG.md:
removed

veronicaz41 commented on apps/confessions-client/pages/index.tsx:
```diff
@@ -0,0 +1,186 @@
+  ? "https://api.pcd-passport.com/semaphore/1"
+  : "http://localhost:3002/semaphore/1";
+
+export default function Web() {
```
Awesome, fixed.
Another problem is error handling, also displaying user friendly errors. Would be great to have that in the example code too.

veronicaz41 commented on apps/confessions-server/src/routing/routes/confessionRoutes.ts:
```diff
@@ -0,0 +1,83 @@
+      const deserialized = await SemaphoreGroupPCDPackage.deserialize(
+        request.proof
+      );
+      const verified = await SemaphoreGroupPCDPackage.verify(deserialized);
```
I realized I shouldn't trust the semaphoreUrl and confession in the request. Fixed.
And yes, we can restrict posting to just the zuzalu group for now.

veronicaz41 commented on apps/confessions-server/src/routing/routes/confessionRoutes.ts:
```diff
@@ -0,0 +1,117 @@
+      const serializedGroup = JSON.parse(json) as SerializedSemaphoreGroup;
+      const group = new Group(1, 16);
+      group.addMembers(serializedGroup.members);
+      if (pcd.proof.proof.merkleTreeRoot !== group.root.toString()) {
```
@ichub I realized that ideally I should be using `pcd.claim` here rather than `proof` (as how I check the signal below). Ideally, third party apps should not be using what's inside proof directly (even inside the integration).
However in this case, it is easier to use the `merrkleTreeRoot` in the proof. And I copied this code from here: https://github.com/proofcarryingdata/zupass/blob/main/packages/passport-interface/src/SemaphoreGroupIntegration.ts#L92-L105.
Wondering how you think about this.

veronicaz41 replied:
cool, will use it once your PR gets merged

veronicaz41 replied:
changed to use `pcd.claim`

veronicaz41 commented:
Addressed PR comments.
Next step: create a log-in flow which feels really solid that would allow users to see confessions only if they are a zuzalu member

gubsheep commented on apps/confessions-server/src/routing/routes/confessionRoutes.ts:
```diff
@@ -0,0 +1,83 @@
+  });
+
+  app.get("/confessions", async (req: Request, res: Response) => {
+    // TODO: only Zuzalu members can see the confessions???
```
yes, if possible a working E2E demonstration of passport auth working cleanly would be great

veronicaz41 replied:
implemented

veronicaz41 commented:
@ichub implemented the login flow. don't have the problem you showed me.
Would be great if you can review it before I move this to the separate repo.

ichub commented on apps/confessions-client/components/Confessions.tsx:
```diff
@@ -0,0 +1,34 @@
+    }
+  
+    (async () => {
+      // TODO: paging
```
we should implement this before we launch, so that we can fire and forget

veronicaz41 replied:
won't get to this until later this week or early next week

ichub commented on apps/confessions-client/components/Confessions.tsx:
```diff
@@ -0,0 +1,34 @@
+    <>
+      <h2>Confessions</h2>
+      {confessions != null && (
+        <pre>{JSON.stringify(confessions, null, 2)}</pre>
```
some rudimentary component for a confession would be nice, rather than `JSON.stringify`

veronicaz41 replied:
won't get to this until later this week or early next week

ichub commented on apps/confessions-client/components/Login.tsx:
```diff
@@ -0,0 +1,88 @@
+import {
```
https://user-images.githubusercontent.com/2636237/229979831-93f6131f-8239-4e0f-9e47-5e37ec08d014.mov

This is really close!

Here are a few things I want changed before we release this though:
- What happens if the user closes the popup window without 'proving' anything? The 'login' button remains disabled. I wonder if 1) some way to re-enable the button when the popup is closed, or 2) if we should disable that button for only a brief moment
- Logging out is instant, as it should be, nice!
- The 'Login'/'Logout' button flickers from 'Login' to 'Logout' if the user is logged in and refreshes the page. I would like it to not flicker, and say 'Logout' immediately if the user is logged in.
- There should be some intermediate feedback state between when the popup delivers a proof to the page, and when the proof is verified. For example, the button could say neither 'Login' nor 'Logout', but 'Verifying'.
- Did you test what happens if the proof that is passed back to the login flow is invalid? That should be handled gracefully as well.

veronicaz41 replied:
I think having an overlay saying "verifying" rather than disable the "login" button, would solve both the following problems.

> What happens if the user closes the popup window without 'proving' anything? The 'login' button remains disabled. I wonder if 1) some way to re-enable the button when the popup is closed, or 2) if we should disable that button for only a brief moment

> There should be some intermediate feedback state between when the popup delivers a proof to the page, and when the proof is verified. For example, the button could say neither 'Login' nor 'Logout', but 'Verifying'.

ichub commented on apps/confessions-client/components/PublishConfession.tsx:
```diff
@@ -0,0 +1,159 @@
+import { PASSPORT_URL, SEMAPHORE_GROUP_URL, requestProofFromPassport } from "../src/util";
+import { postConfession } from "../src/api";
+
+export function PublishConfession({
```
Add a comment to this component.

veronicaz41 replied:
added

ichub commented on apps/confessions-client/components/Login.tsx:
```diff
@@ -0,0 +1,88 @@
+import { PASSPORT_URL, SEMAPHORE_GROUP_URL, requestProofFromPassport } from "../src/util";
+import { login } from "../src/api";
+
+export function Login({
```
Add a comment to this component.

veronicaz41 replied:
added

ichub commented on apps/confessions-client/components/PublishConfession.tsx:
```diff
@@ -0,0 +1,159 @@
+      >
+        Publish
+      </button>
+      {proof != null && (
+        <>
+          <h3>Got Zuzalu Member Confession Proof from Passport</h3>
+          <pre>{JSON.stringify(proof, null, 2)}</pre>
+          {valid === undefined && <p>❓ Proof verifying</p>}
+          {valid === false && <p>❌ Proof is invalid</p>}
+          {valid === true && <p>✅ Proof is valid</p>}
+        </>
+      )}
```
for the version that we release as an example, we should hide this

veronicaz41 replied:
deleted this
added a console.log(proof), thought might still be useful for developers as this is a demo app

ichub commented on apps/confessions-client/package.json:
```diff
@@ -0,0 +1,36 @@
+    "@semaphore-protocol/group": "^3.2.3",
+    "@semaphore-protocol/identity": "^3.2.3",
+    "next": "^13.1.1",
+    "@pcd/passport-interface": "*",
+    "@pcd/pcd-types": "*",
+    "@pcd/semaphore-group-pcd": "*",
+    "@pcd/semaphore-identity-pcd": "*",
```
(Let's not forget this - I published new versions - all the packages are now at `0.1.0`)

veronicaz41 replied:
changed to use 0.1.0 for all pcd packages

ichub commented on apps/confessions-client/pages/index.tsx:
```diff
@@ -0,0 +1,61 @@
+  const [accessToken, setAccessToken] = useState<string | null>(null);
+  const [newConfession, setNewConfession] = useState<string | undefined>();
+
+  useEffect(() => {
+    if (accessToken) return;
+
+    const token = window.localStorage.getItem("access_token");
+    setAccessToken(token);
+  }, [accessToken]);
+
+  const handleAccessToken = (token: string | null) => {
+    setAccessToken(token)
+    if (!token) {
+      window.localStorage.removeItem("access_token");
+    } else {
+      window.localStorage.setItem("access_token", token!);
+    }
+  }
```
I think we should extract this hook/state, along with the logic that determines whether to show the user a login or logout button into its own component called `LoginLogout`. To make it not flicker (the flickering i pointed out earlier), you could change

```ts
const [accessToken, setAccessToken] = useState<string | null>(null);
```

to 
```ts
const [accessToken, setAccessToken] = useState<string | null>(() => window.localStorage.getItem("access_token") ?? null);
```

since localstorage is a synchronous API in web (thankfully)

ichub commented on apps/confessions-client/pages/index.tsx:
```diff
@@ -0,0 +1,61 @@
+    setAccessToken(token);
+  }, [accessToken]);
+
+  const handleAccessToken = (token: string | null) => {
```
a better name might be `updateAccessToken`

veronicaz41 replied:
changed

ichub commented on apps/confessions-client/pages/popup.tsx:
```diff
@@ -0,0 +1,33 @@
+ * a PCD from the passport. It uses the window messaging API to communicate
+ * the PCD it received back to the requesting tab.
+ */
+export default function AuthPopup() {
```
this code is now duplicated in a bunch of 3rd party apps, so we should extract it. not relevant for this pr just noting it. here's an issue: https://github.com/proofcarryingdata/zupass/issues/131

ichub commented on apps/confessions-server/.env.example:
```diff
@@ -0,0 +1,19 @@
+# to be able to create URLs that point to itself.
+CONFESSIONS_SERVER_URL="http://localhost:3005"
+
+ACCESS_TOKEN_SECRET=
```
add a comment regarding what, if anything, this should be set to in development

veronicaz41 replied:
added

ichub commented on apps/confessions-server/.env.example:
```diff
@@ -0,0 +1,19 @@
+# Prisma supports the native connection string format for PostgreSQL, MySQL, SQLite, SQL Server, MongoDB and CockroachDB.
+# See the documentation for all the connection string options: https://pris.ly/d/connection-strings
+
+DATABASE_URL="postgresql://johndoe:randompassword@localhost:5432/mydb?schema=public"
```
add documentation regarding how this database should be initialized for it to be accessible by the app

ichub commented on apps/confessions-server/src/main.ts:
```diff
@@ -0,0 +1,14 @@
+import { IS_PROD } from "./util/isProd";
+
+const dotEnvPath = IS_PROD
+  ? `/etc/secrets/.env`
```
once you deploy this to render.com, be sure to make the deployment configuration create a `.env` file, and this directory is where it goes by default

ichub commented on apps/confessions-server/src/routing/routes/confessionRoutes.ts:
```diff
@@ -0,0 +1,152 @@
+  // this endpoint verifies the proof is valid,
+  // and the proof claim matches the semaphoreGroupUrl and the confession in the request
+  // before saving the confession to the database.
+  app.post("/confessions", async (req: Request, res: Response, next: NextFunction) => {
```
a better name for this route could be `"/new-confession"`

ichub replied:
Another thing is that technically this route could just use the `jwt` to authenticate, right? Having both authentication mechanisms in this example app may be confusing, but I'll leave it to you to decide whether it makes sense.

ichub replied:
I think another issue of verifying this proof for posting a confession is that if the proof was generated a while ago, then the proof may be outdated, since new ppl have been added to the group since then. that's not ideal.

veronicaz41 replied:
Changed the route name.

Technically this route could use the `jwt` to authenticate. Still want to keep it to use pcd for demonstration purposes, since this is a demo app for developers. The pcd will verify the user belongs to the zuzalu group and also made the confession (a signal).
I split the routes file to `authedRoutes` which contains `/login` and get `/confessions`. It shows how authenticate with jwt would work. And `pcdRoutes` which contains this `/new-confession`. It shows how to pcd can be used.

veronicaz41 replied:
> I think another issue of verifying this proof for posting a confession is that if the proof was generated a while ago, then the proof may be outdated, since new ppl have been added to the group since then. that's not ideal.

This is definitely an issue.
However, it is a general issue which applies to /login too, and any pcd using the semaphore group proof in general.
We could snapshot the group merkletree whenever there's changes to the group. However, if it happens frequently, it could be problematic.
How do we solve the problem for the zuzalu passport? When a new zuzalu is added, do we update the content fetched from the semaphore_group_url? @ichub

ichub commented on apps/confessions-server/src/routing/routes/confessionRoutes.ts:
```diff
@@ -0,0 +1,152 @@
+  signal?: string
+): Promise<Error | null> {
+  // only allow Zuzalu group for now
+  if (semaphoreGroupUrl !== SEMAPHORE_GROUP_URL) {
```
maybe we could make `SEMAPHORE_GROUP_URL` load from an environment variable, and document that environment variable. this would make it much easier for a newbie dev to clone this repo and make it work with some other arbitrary semaphore group.

veronicaz41 replied:
made SEMAPHORE_GROUP_URL to load from env var, also added the documentation

ichub commented on apps/confessions-server/src/services/types.ts:
```diff
@@ -0,0 +1,13 @@
+  (context: ApplicationContext): void;
+}
+
+export interface Metric {
+  (context: ApplicationContext): void;
+}
+
+export enum MetricName {
+  GITHUB_RATE_LIMIT = "GITHUB_RATE_LIMIT",
+}
```
can delete these

veronicaz41 replied:
deleted

ichub commented on apps/confessions-server/src/util/auth.ts:
```diff
@@ -0,0 +1,36 @@
+import { verify, JwtPayload } from "jsonwebtoken";
+import { IS_PROD } from "./isProd";
+
+export const ACCESS_TOKEN_SECRET = process.env.ACCESS_TOKEN_SECRET || "secrect";
```
`secrect` lol

veronicaz41 replied:
the 'r's and 'c's do not work very well on my keyboard of this very old laptop. it is a miracle what I type compiles most of the time.

ichub commented on apps/confessions-server/tsconfig.json:
```diff
@@ -0,0 +1,12 @@
+    "typeRoots": [
+      "src/util/declarations/",
+      "./node_modules/@types/",
+      "../../node_modules/@types/"
```
you'll need to update this path when you extract to another repository

veronicaz41 commented:
Moving the code to the separate repo at this commit
Will address the remaining comments in the separate repo

veronicaz41 commented:
@ichub I think you need to add `server.json` here: https://github.com/proofcarryingdata/zupass/blob/main/packages/tsconfig/package.json#L6-L10. Otherwise, I'm seeing `TSError: ⨯ Unable to compile TypeScript:
confessions-server:dev: error TS6053: File '@pcd/tsconfig/server.json' not found.` in my separate repo.

ichub commented:
> @ichub I think you need to add `server.json` here: https://github.com/proofcarryingdata/zupass/blob/main/packages/tsconfig/package.json#L6-L10. Otherwise, I'm seeing `TSError: ⨯ Unable to compile TypeScript: confessions-server:dev: error TS6053: File '@pcd/tsconfig/server.json' not found.` in my separate repo.

for now you should just merge whatever tsconfig you have in that repo with `server.json` in this repo. separately, we should definitely be publishing those configs on npm. here's an issue for that: https://github.com/proofcarryingdata/zupass/issues/138

<img width="737" alt="Screenshot 2023-04-06 at 9 13 52 AM" src="https://user-images.githubusercontent.com/2636237/230255203-30972508-22c8-43cd-b16e-44016dc3826f.png">

veronicaz41 commented:
The code for this PR is here now: https://github.com/proofcarryingdata/zuzalu-confessions.
Will continue to work on that repo.

ichub commented:
closing since we've migrated this app to its own repo: https://github.com/proofcarryingdata/zuzalu-confessions
